pkg/operator/controller/machineconfig/crio.go

```diff
@@ -0,0 +1,201 @@
+limitations under the License.
+*/
+
+package machineconfigcontroller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntime "sigs.k8s.io/controller-runtime"
+
+	ignutil "github.com/coreos/ignition/v2/config/util"
+	igntypes "github.com/coreos/ignition/v2/config/v3_2/types"
+	mcv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+)
+
+const (
+	// CrioProfilingConfigName is the name CRI-O MachineConfig CR
+	CrioProfilingConfigName = "99-crio-profiling"
+
+	// CrioUnixSocketConfFile is the name of the CRI-O config file
+	CrioUnixSocketConfFile = "10-mco-profile-unix-socket.conf"
+
+	// CriounixSocketConfData contains the configuration required
+	// for enabling CRI-O profiling
+	CriounixSocketConfData = `[Service]
+Environment="ENABLE_PROFILE_UNIX_SOCKET=true"`
+)
+
+// ensureCrioProfConfigExists checks if CRI-O MachineConfig CR for
+// enabling profiling exists, if not creates the resource
+func (r *MachineconfigReconciler) ensureCrioProfConfigExists(ctx context.Context) (*mcv1.MachineConfig, bool, error) {
+	namespace := types.NamespacedName{Name: CrioProfilingConfigName}
+	criomc, exists, err := r.fetchCrioProfileConfig(ctx, namespace)
+	if err != nil {
+		return nil, false, err
+	}
+	if !exists {
+		if err := r.createCrioProfileConfig(ctx); err != nil {
+			return nil, false, err
+		}
+
+		criomc, found, err := r.fetchCrioProfileConfig(ctx, namespace)
+		if err != nil || !found {
+			return nil, false, fmt.Errorf("failed to fetch just created crio config: %v", err)
+		}
+
+		return criomc, true, nil
+	}
+	return criomc, false, nil
+}
+
+// ensureCrioProfConfigNotExists checks if CRI-O MachineConfig CR for
+// enabling profiling exists, if exists delete the resource
+func (r *MachineconfigReconciler) ensureCrioProfConfigNotExists(ctx context.Context) (bool, error) {
+	namespace := types.NamespacedName{Name: CrioProfilingConfigName}
+	criomc, exists, err := r.fetchCrioProfileConfig(ctx, namespace)
+	if err != nil {
+		return false, err
+	}
+	if exists {
+		if err := r.deleteCrioProfileConfig(ctx, criomc); err != nil {
+			return false, err
+		}
+
+		_, exists, err = r.fetchCrioProfileConfig(ctx, namespace)
+		if err != nil || exists {
+			return false, fmt.Errorf("failed to delete crio profiling config: %v", err)
+		}
+
+		return true, nil
+	}
+	return false, nil
+}
+
+// fetchCrioProfileConfig is for fetching the CRI-O MC CR created
+// by this controller for enabling profiling
+func (r *MachineconfigReconciler) fetchCrioProfileConfig(ctx context.Context, namespace types.NamespacedName) (*mcv1.MachineConfig, bool, error) {
+	criomc := &mcv1.MachineConfig{}
+
+	if err := r.Get(ctx, namespace, criomc); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, false, nil
```

Comment:
I don't understand the semantics here: why the `bool` value?
This is exactly what `errors.IsNotFound(err)` is for.

Reply:
If fetching a resource fails with NotFound error it's not considered as error, and the bool value is used for indicating the same, and the caller checks whether err and if not does it exist. Thought to use extra return value to blindly return error if any.